### PR TITLE
tcp_proxy: explicitly supporting upstream-write-first semantics in the new pool

### PR DESCRIPTION
**Pull request by alyssawilk**
Additional Description:
Both the old and the new TCP proxy pools register for "unexpected data" and disconnect when receiving it.  It appears that generally the timing works out in the old pool that the connection gets assigned right away, but the new pool doesn't pass the race deterministically.  Fixing this by read-disabling, so data will be associated with the downstream connection as it is associated.
Risk Level: Medium (new pool is still off by default, but affects ALPN as well)
Testing: new integration tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/http/mixed_conn_pool.cc [hunk 1]
```diff
@@ -50,6 +50,7 @@ void HttpConnPoolImplMixed::onConnected(Envoy::ConnectionPool::ActiveClient& cli
 
   Upstream::Host::CreateConnectionData data{std::move(tcp_client->connection_),
                                             client.real_host_description_};
+  data.connection_->readDisable(false);
```
Comment (ggreenway) [inline source/common/http/mixed_conn_pool.cc:53]: Why is this needed? I haven't followed development of mixed_conn_pool; can this take a connection from the tcp-proxy pool which was read-disabled?

File: source/common/tcp/conn_pool.h [hunk 2]
```diff
@@ -94,6 +94,18 @@ class ActiveTcpClient : public Envoy::ConnectionPool::ActiveClient {
   void onAboveWriteBufferHighWatermark() override { callbacks_->onAboveWriteBufferHighWatermark(); }
   void onBelowWriteBufferLowWatermark() override { callbacks_->onBelowWriteBufferLowWatermark(); }
 
+  // Undo the readDisable done in onEvent(Connected) - now that there is an associated connection,
+  // drain any data.
+  void readEnableIfNew() {
+    if (!associated_before_) {
```
Comment (ggreenway) [inline source/common/tcp/conn_pool.h:100]: What path leads to this being called more than once on a connection?

File: source/common/http/mixed_conn_pool.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/http/mixed_conn_pool.cc:56]: yeah, the ALPN pool creates a raw TCP connection, then slaps on an H1/H2 codec once we have TLS handshake results.  I'll add a comment to clarify

File: source/common/tcp/conn_pool.h (hunk 2 above)
Comment (alyssawilk) [reply source/common/tcp/conn_pool.h:100]: for Envoy, it never is.  AFIK we allow connection reuse and added it for someone who uses it in-house, and I don't want to break things for them.

Comment (ggreenway) [reply source/common/tcp/conn_pool.h:100]: Can you add a comment to that effect?

**Review by ggreenway — APPROVED** [review]

**Comment by ggreenway** [comment]
Mac CI failed on tcp_proxy_integration_test. Is it a flake related to this PR, or an unrelated flake?

/wait-any

**Review by ggreenway — APPROVED** [review]

**Comment by alyssawilk** [comment]
Merged! props to @rgs1 for all his hard work tracking this one down :-D
